pkg/util/acrtoken/acrtoken_test.go

```diff
@@ -68,3 +70,134 @@ func TestEnsureTokenAndPassword(t *testing.T) {
+			tokenPasswordProperties: []mgmtcontainerregistry.TokenPassword{
+				{
+					Name:         mgmtcontainerregistry.TokenPasswordNamePassword1,
+					CreationTime: &date.Time{Time: time.Date(2022, time.April, 4, 0, 0, 0, 0, time.UTC)},
+				},
+				{
+					Name:         mgmtcontainerregistry.TokenPasswordNamePassword2,
+					CreationTime: &date.Time{Time: time.Date(2022, time.May, 20, 0, 0, 0, 0, time.UTC)},
+				},
+			},
+			wantRenewalName: mgmtcontainerregistry.TokenPasswordNamePassword1,
+			wantPassword:    "foo",
+		},
+		{
+			name: "renews password2 when it is the oldest password",
+			tokenPasswordProperties: []mgmtcontainerregistry.TokenPassword{
+				{
+					Name:         mgmtcontainerregistry.TokenPasswordNamePassword1,
+					CreationTime: &date.Time{Time: time.Date(2022, time.May, 20, 0, 0, 0, 0, time.UTC)},
+				},
+				{
+					Name:         mgmtcontainerregistry.TokenPasswordNamePassword2,
+					CreationTime: &date.Time{Time: time.Date(2022, time.April, 4, 0, 0, 0, 0, time.UTC)},
+				},
+			},
+			wantRenewalName: mgmtcontainerregistry.TokenPasswordNamePassword2,
+			wantPassword:    "bar",
+		},
+	}
+
+	username := "aro-12345"
+	resourceID := "/subscriptions/93aeba23-2f76-4307-be82-02921df010cf/resourceGroups/global/providers/Microsoft.ContainerRegistry/registries/arointsvc"
+	ctx := context.Background()
+	controller := gomock.NewController(t)
+
+	env := mock_env.NewMockInterface(controller)
+	env.EXPECT().ACRResourceID().AnyTimes().Return(resourceID)
+	tokens := mock_containerregistry.NewMockTokensClient(controller)
+	registries := mock_containerregistry.NewMockRegistriesClient(controller)
+	r, err := azure.ParseResourceID(env.ACRResourceID())
+	if err != nil {
+		t.Fatal(err)
+	}
+	credentialResult := mgmtcontainerregistry.GenerateCredentialsResult{
+		Passwords: &[]mgmtcontainerregistry.TokenPassword{
+			{
+				Name:  mgmtcontainerregistry.TokenPasswordNamePassword1,
+				Value: to.StringPtr("foo"),
+			},
+			{
+				Name:  mgmtcontainerregistry.TokenPasswordNamePassword2,
+				Value: to.StringPtr("bar"),
+			},
+		},
+	}
+	registryProfile := api.RegistryProfile{
+		Username: username,
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeTokenProperties := mgmtcontainerregistry.TokenProperties{
+				Credentials: &mgmtcontainerregistry.TokenCredentialsProperties{
+					Passwords: &tt.tokenPasswordProperties,
+				},
+			}
+			tokens.EXPECT().GetTokenProperties(ctx, "global", "arointsvc", username).Return(fakeTokenProperties, nil)
+
+			expectedCredentialParameters := mgmtcontainerregistry.GenerateCredentialsParameters{
+				TokenID: to.StringPtr(resourceID + "/tokens/" + username),
+				Name:    tt.wantRenewalName,
+			}
+			registries.EXPECT().GenerateCredentials(ctx, "global", "arointsvc", expectedCredentialParameters).
+				Return(credentialResult, nil)
+
+			m := &manager{
+				env: env,
+				r:   r,
+
+				registries: registries,
+				tokens:     tokens,
+			}
```

Comment:
alternatively, writing "constructor" functions for `registries`, `tokens`, and `env`, then passing them to the manager would also add clarity, just something that abstracts this away, and puts all the EXPECTS in one place would really help the legibility here.

Reply:
I've extracted some of the setup logic out as best as I can. Part of the problem is once the mocks are passed to the manager they are considered to be the interfaces in the manager, so they no longer have access to the `EXPECT` function that gomock gives them. Those `EXPECT`s are being used both to ensure that the Azure SDK functions are being called with the correct parameters and to set up the response that is being used in the rest of the function. If we wanted to reduce the amount being done by a single function we could _probably_ break some of the behavior out into sub-functions, but I didn't do that while writing because I didn't have any concerns about reusability.